File: graphs/reducers/git-commit.js
// Commit all changes inside a worktree (a side effect: gate it). Returns the diff stat and commit hash.
// input: { worktree, message, author? }
import { execFileSync } from "node:child_process";
import path from "node:path";

export default async function gitCommit(input, args, ctx) {
  const cwd = path.resolve(String(input.worktree));
  const git = (...a) => execFileSync("git", a, { cwd, encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] }).trim();
  const status = git("status", "--porcelain");
  if (!status) return { committed: false, reason: "no changes in worktree", commit: null, stat: "" };
  git("add", "-A");
  const message = String(input.message ?? args.message ?? "gren: automated change");
  const trailer = `\n\nCo-Authored-By: ${String(input.author ?? args.author ?? "gren <noreply@example.com>")}`;
  git("-c", "user.name=gren", "-c", "user.email=gren@localhost", "commit", "-q", "-m", message + trailer);
  const commit = git("rev-parse", "HEAD");
  const stat = git("show", "--stat", "--oneline", "HEAD");
  ctx.log(`committed ${commit.slice(0, 8)} in ${cwd}`);
  return { committed: true, commit, branch: git("rev-parse", "--abbrev-ref", "HEAD"), stat, files_changed: status.split("\n").length };
}

File: graphs/reducers/git-commit.py
# Commit all changes inside a worktree (a side effect: gate it). Returns the diff stat and commit hash.
# input: { worktree, message, author? }
import os
import subprocess


def _git(cwd: str, *a: str) -> str:
    return subprocess.run(["git", *a], cwd=cwd, capture_output=True, text=True, encoding="utf-8", errors="replace", check=True).stdout.strip()


def reduce(input, args, ctx):
    cwd = os.path.abspath(str(input.get("worktree")))
    status = _git(cwd, "status", "--porcelain")
    if not status:
        return {"committed": False, "reason": "no changes in worktree", "commit": None, "stat": ""}
    _git(cwd, "add", "-A")
    message = str(input.get("message") or args.get("message") or "gren: automated change")
    trailer = f"\n\nCo-Authored-By: {input.get('author') or args.get('author') or 'gren <noreply@example.com>'}"
    _git(cwd, "-c", "user.name=gren", "-c", "user.email=gren@localhost", "commit", "-q", "-m", message + trailer)
    commit = _git(cwd, "rev-parse", "HEAD")
    ctx.log(f"committed {commit[:8]} in {cwd}")
    return {"committed": True, "commit": commit, "branch": _git(cwd, "rev-parse", "--abbrev-ref", "HEAD"), "stat": _git(cwd, "show", "--stat", "--oneline", "HEAD"), "files_changed": len(status.splitlines())}
